<a href="https://colab.research.google.com/github/francomanca93/fundamentos-de-estadistica-con-python/blob/analisis-exploratorio/2.%20Caracterizar%20informaci%C3%B3n%20a%20traves%20del%20an%C3%A1lisis%20exploratorio/3_Medidas_de_tendencia_central.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Medidas de tendencia central

Las [medidas de tendencia central](https://es.wikipedia.org/wiki/Medidas_de_tendencia_central) son valores principales para caracterizar variables aleatorias.

La medida de tendencia central es un número situado hacia el centro de la distribución de los valores de una serie de observaciones (medidas), en la que se encuentra ubicado el conjunto de los datos.

Es importante que no todas las medidas de tendencia son aplicables a los dos tipos de variables, numéricas y categóricas, y que también algunas de ellas son más susceptibles a los valores extremos (outliers)

**Medidas de tendencia**

- Media
  - [Aritmética]((https://es.wikipedia.org/wiki/Media_aritm%C3%A9tica))
  - [Geométrica](https://es.wikipedia.org/wiki/Media_geom%C3%A9trica)
  - [Armónica](https://es.wikipedia.org/wiki/Media_arm%C3%B3nica)
  - [Ponderada](https://es.wikipedia.org/wiki/Media_ponderada)
- [Mediana](https://es.wikipedia.org/wiki/Mediana_(estad%C3%ADstica))
- [Moda](https://es.wikipedia.org/wiki/Moda_(estad%C3%ADstica))
- [Error típico o desviación estándar](https://es.wikipedia.org/wiki/Desviaci%C3%B3n_t%C3%ADpica)

**Criterios de aplicación**

- No todas las medidas de tendencia central son aplicables a las variables numericas y categoricas, o sea, hay que tener en cuenta el tipo de variable a la hora de utilizarlas. 

- Alguna de ellas son mas suceptibles que otras a los valores extremos.


Importamos librerias que vamos a utilizar

In [1]:
import numpy as np
import pandas as pd
import scipy
import scipy.stats

Montamos drive para acceder a la base de datos con la cual vamos a trabajar

In [2]:
#Montamos Drive
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [3]:
ls '/content/drive/My Drive/Colab Notebooks/fundamentos-de-estadistica-con-python/datos/'

bicicletas-compartidas.csv  datos_profesora.py  __pycache__/


In [4]:
cd '/content/drive/My Drive/Colab Notebooks/fundamentos-de-estadistica-con-python/datos/'

/content/drive/My Drive/Colab Notebooks/fundamentos-de-estadistica-con-python/datos


Importamos la base de datos y empezamos a trabajar con ella

In [7]:
df = pd.read_csv('bicicletas-compartidas.csv')
df.columns

Index(['fecha', 'bicis-compartidas', 'temp-obs', 'sens-temp', 'hum', 'viento',
       'codigo-clima', 'festivo', 'findesemana', 'cuartil-ano'],
      dtype='object')

Vamos a definir la variable con la cual vamos a generar el analisis. Lo haremos sobre las bicicleras compartidas.

In [8]:
y = df['bicis-compartidas'].values
y

array([182, 138, 134, ..., 337, 224, 139])

Vemos que **y** es un arreglo numérico. Ahora vamos a limpiar este arreglo de posibles 0 que tenga, ya que numpy no nos permitira hacer calculo con estos valores.

In [9]:
y = np.where(y == 0, 1, y)

In [10]:
np.min(y)

1

In [11]:
np.max(y)

7860

## Promedio aritmetico o media aritmetica

La [media aritmética](https://es.wikipedia.org/wiki/Media_aritm%C3%A9tica) también llamada promedio o media, de un conjunto finito de números es el valor característico de una serie de datos cuantitativos, objeto de estudio que parte del principio de la esperanza matemática o valor esperado, se obtiene a partir de la suma de todos sus valores dividida entre el número de sumandos.

![Media aritmetica](https://wikimedia.org/api/rest_v1/media/math/render/svg/45369dd1e80668d3f6df7233fe4e2e0361d5c851)

In [12]:
np.mean(y)

1143.1016997817849

In [13]:
np.sum(y)/len(y)

1143.1016997817849

Vemos que la media aritmética es de 1143 bicicletas compartidas

Documentacion oficial de scipy mstats:

https://docs.scipy.org/doc/scipy/reference/stats.mstats.html

## Promedio geometrica o media geométrica

La [media geométrica](https://es.wikipedia.org/wiki/Media_geom%C3%A9trica) de una cantidad arbitraria de números (por decir n números) es la raíz n-ésima del producto de todos los números; es recomendada para datos de progresión geométrica, para promediar razones, interés compuesto y números índice

![Media Geometrica](https://wikimedia.org/api/rest_v1/media/math/render/svg/ec635835ec23391e636a8dea960c2399c95b19aa)

Frecuentemente se usa una media geométrica cuando se comparan diferentes aspectos, cuyos rendimientos tienen unidades de medida en diferentes rangos numéricos. 

Por ejemplo, la media geométrica puede dar un valor serio para comparar dos empresas que tienen una calificación entre 0 a 5 por su sostenibilidad ambiental, y una calificación entre 0 a 100 por su viabilidad financiera. Si se usara la media aritmética en lugar de la media geométrica, la viabilidad financiera tendría mayor peso porque su rango numérico es mayor. Es decir, un pequeño cambio porcentual en la calificación financiera (por ejemplo, pasar de 80 a 90) haría una diferencia mucho mayor en la media aritmética que un gran cambio porcentual en la sostenibilidad ambiental (por ejemplo, pasar de 2 a 5). El uso de la media geométrica normaliza los valores de rango diferente, lo que significa que un cambio de porcentaje dado en cualquiera de las propiedades tiene el mismo efecto en la media geométrica. Entonces, un cambio del 20% en la sostenibilidad ambiental de 4 a 4.8 tiene el mismo efecto en la media geométrica que un cambio del 20% en la viabilidad financiera de 60 a 72. 

Esto lo hacemos de la siguiente manera

In [15]:
scipy.stats.mstats.gmean(y)

623.2472122430046

## Promedio armonico o media armonica

La [media armónica](https://es.wikipedia.org/wiki/Media_arm%C3%B3nica) (designada usualmente mediante H) de una cantidad finita de números es igual al recíproco, o inverso, de la media aritmética de los recíprocos de dichos valores y es recomendada para promediar velocidades. 

![Media armonica](https://wikimedia.org/api/rest_v1/media/math/render/svg/31bb15ba938382149ac4d2ecda7479effc6ae4d6)


In [14]:
scipy.stats.mstats.hmean(y)

241.65180305136826

## Mediana

La [mediana](https://es.wikipedia.org/wiki/Mediana_(estad%C3%ADstica)) representa el valor de la variable de posición central en un conjunto de datos ordenados, en otras palabras el valor que divide a los datos en 2 partes. Coincide con el percentil 50% o 50% de los datos.


In [17]:
np.median(y)

844.0

## Moda

La [moda](https://es.wikipedia.org/wiki/Moda_(estad%C3%ADstica)) es el valor con mayor frecuencia en una de las distribuciones de datos.

In [18]:
# moda = valor de y con la maxima frecuencia

moda = np.nan # defino la variable moda

valores, conteo_valores = np.unique(y, return_counts=True) # Retorna array con valores unicos del array que le pase.

pos = np.argmax(conteo_valores) # Del conteo de valores retorname el valor máximo. Este será la posición
moda = valores[pos] # Con la posicion anterior encuentro el valor que mas se repite.

In [20]:
valores

array([   1,    9,   10, ..., 7208, 7531, 7860])

In [21]:
conteo_valores

array([1, 1, 1, ..., 1, 1, 1])

In [22]:
pos

38

In [23]:
moda

46

El valor que mas se repite es 46, 46 bicicletas compartidas.

Con esto finalizamos las medidas de tendencia central, en el siguiente gráfico podemos ver una comparativa entre estas medidas, obtenida de wikipedia

![Comparacion de medidas de tendencia central](https://upload.wikimedia.org/wikipedia/commons/thumb/3/33/Visualisation_mode_median_mean.svg/langes-320px-Visualisation_mode_median_mean.svg.png)

## Desviación estandar

La [desviación estandar](https://es.wikipedia.org/wiki/Desviaci%C3%B3n_t%C3%ADpica) es una medida que se utiliza para cuantificar la variación o la dispersión de un conjunto de datos numéricos.

Una desviación estándar baja indica que la mayor parte de los datos de una muestra tienden a estar agrupados cerca de su media (también denominada el valor esperado), mientras que una desviación estándar alta indica que los datos se extienden sobre un rango de valores más amplio. 

![Desviación estandar](https://upload.wikimedia.org/wikipedia/commons/thumb/8/8c/Standard_deviation_diagram.svg/1280px-Standard_deviation_diagram.svg.png)

Una gráfica de la distribución normal (o curva en forma de campana, o curva de Gauss), donde cada banda tiene un ancho de una vez la desviación estándar

In [24]:
np.std(y)

1085.0768508213835

## Alteraciones

Podemos ver que los valores extremos alteran las medidas de tendencia centrla. Veamos esto con un ejemplo:

In [27]:
y_alterado = y.copy()
y_alterado[y_alterado == max(y_alterado)] = 10000000

In [28]:
print(np.mean(y))
print(np.mean(y_alterado))

1143.1016997817849
1716.9009417709888


In [29]:
print(np.median(y))
print(np.median(y_alterado))

844.0
844.0


Veamos lo anterior con un gráfico bastante representativo.
![alteraciones](https://imgur.com/L6drVeI.png)